libminifi/include/utils/FlatMap.h

```diff
@@ -0,0 +1,211 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef LIBMINIFI_INCLUDE_UTILS_FLATMAP_H_
+#define LIBMINIFI_INCLUDE_UTILS_FLATMAP_H_
+
+#include <tuple>
+#include <functional>
+#include <vector>
+#include <utility>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace utils {
+
+template<typename K, typename V>
+class FlatMap{
+ public:
+  using value_type = std::pair<K, V>;
```

Comment:
I had no intention of making `FlatMap` adhere to any of the named requirements (that is why it does not), it seems like `AssociativeContainer` expects an ordering, should we go for `Container`?

(frankly I ~~despise~~ dislike named requirements as they just further push the responsibilities to the programmer, which IMO should belong to the compiler, e.g. `Container` expects a `_.cbegin()` with constant complexity, so what happens if it is not, surely then the compiler rejects the program if we use the type where a `Container` is expected... in our dreams, mind you this is not unique to C++)

Reply:
My point is that if you write a container, then it should behave like a canonical container so that it can work as a drop-in replacement for at least some containers. Named requirements (i.e. concepts) just clarify the properties of a canonical container so that implementors have guidelines to follow and users have an interface to code against. They are just the generic programming equivalent of an interface, with less dependencies and less opportunity for the compiler to check whether the requirements are satisfied. You can not implement `java.util.Collection` without implementing `isEmpty`, but the compiler can not yet enforce that your Collection implementation has `empty`.

I think the closer the satisfied concepts are to the reality, the better. Ideally, this would satisfy AssociativeContainer and iterators would satisfy RandomAccessIterator, but I'm OK with postponing the implementation of stricter requirements until they are actually needed.

Reply:
Alternatively, we can leave it as non-generic and hide the implementation as an implementation detail of `FlowFile`, but we would lose software capital by choosing that route.

Reply:
it can check the existence and signature of methods, although until C++20 we need something like `std::is_collection`, what it cannot do and will not do (at least not c++) is check semantic requirements (like the complexity requirement I mentioned), and worse yet we cannot event express the requirement in code, so if we want to check that a type satisfies the requirements of `Collection` we must consult the C++ specification

Reply:
I'm not against making `FlatMap` generic enough (by making it fulfil `Container`) for possible future

Reply:
This seems to be done, closing